# Desafío Telecom X parte 2

## Objetivos del desafío
- Preparar los datos para el modelado (tratamiento, codificación, normalización).
- Realizar análisis de correlación y selección de variables.
- Entrenar dos o más modelos de clasificación.
- Evaluar el rendimiento de los modelos con métricas.
- Interpretar los resultados, incluyendo la importancia de las variables.
- Crear una conclusión estratégica señalando los principales factores que influyen en la cancelación.

## Qué se pondrá en práctica
- Preprocesamiento de datos para Machine Learning
- Construcción y evaluación de modelos predictivos
- Interpretación de resultados y entrega de insights
- Comunicación técnica con enfoque estratégico

# Eliminación de columnas irrelevantes

In [45]:
import pandas as pd

In [46]:
datos = pd.read_csv('data/datos_tratados.csv')
datos.head()

customerID  Churn  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  0002-ORFBO      0  Female              0        1           1       9   
1  0003-MKNFE      0    Male              0        0           0       9   
2  0004-TLHLJ      1    Male              0        0           0       4   
3  0011-IGKFF      1    Male              1        1           0      13   
4  0013-EXCHZ      1  Female              1        1           0       3   

   PhoneService MultipleLines InternetService  ... PhoneService_bin  \
0             1            No             DSL  ...                0   
1             1           Yes             DSL  ...                0   
2             1            No     Fiber optic  ...                0   
3             1            No     Fiber optic  ...                0   
4             1            No     Fiber optic  ...                0   

  MultipleLines_bin InternetService_bin OnlineSecurity_bin OnlineBackup_bin  \
0                 0                   0                  0                1   
1                 1                   0                  0                0   
2                 0                   0                  0                0   
3                 0                   0                  0                1   
4                 0                   0                  0                0   

  DeviceProtection_bin TechSupport_bin  StreamingTV_bin StreamingMovies_bin  \
0                    0               1                1                   0   
1                    0               0                0                   1   
2                    1               0                0                   0   
3                    1               0                1                   1   
4                    0               1                1                   0   

   servicios_contratados  
0                      3  
1                      2  
2                      1  
3                      4  
4                      2  

[5 rows x 33 columns]

In [47]:
datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 33 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   customerID             7043 non-null   object 
 1   Churn                  7043 non-null   int64  
 2   gender                 7043 non-null   object 
 3   SeniorCitizen          7043 non-null   int64  
 4   Partner                7043 non-null   int64  
 5   Dependents             7043 non-null   int64  
 6   tenure                 7043 non-null   int64  
 7   PhoneService           7043 non-null   int64  
 8   MultipleLines          7043 non-null   object 
 9   InternetService        7043 non-null   object 
 10  OnlineSecurity         7043 non-null   object 
 11  OnlineBackup           7043 non-null   object 
 12  DeviceProtection       7043 non-null   object 
 13  TechSupport            7043 non-null   object 
 14  StreamingTV            7043 non-null   object 
 15  Stre

In [48]:
datos.describe()

Churn  SeniorCitizen      Partner   Dependents       tenure  \
count  7043.000000    7043.000000  7043.000000  7043.000000  7043.000000   
mean      0.265370       0.162147     0.483033     0.299588    32.371149   
std       0.441561       0.368612     0.499748     0.458110    24.559481   
min       0.000000       0.000000     0.000000     0.000000     0.000000   
25%       0.000000       0.000000     0.000000     0.000000     9.000000   
50%       0.000000       0.000000     0.000000     0.000000    29.000000   
75%       1.000000       0.000000     1.000000     1.000000    55.000000   
max       1.000000       1.000000     1.000000     1.000000    72.000000   

       PhoneService  PaperlessBilling  Charges.Monthly  Charges.Total  \
count   7043.000000       7043.000000      7043.000000    7043.000000   
mean       0.903166          0.592219        64.761692    2279.734304   
std        0.295752          0.491457        30.090047    2266.794470   
min        0.000000          0.000000        18.250000       0.000000   
25%        1.000000          0.000000        35.500000     398.550000   
50%        1.000000          1.000000        70.350000    1394.550000   
75%        1.000000          1.000000        89.850000    3786.600000   
max        1.000000          1.000000       118.750000    8684.800000   

       Cuentas_Diarias  ...  PhoneService_bin  MultipleLines_bin  \
count      7043.000000  ...            7043.0        7043.000000   
mean          2.158675  ...               0.0           0.421837   
std           1.003088  ...               0.0           0.493888   
min           0.610000  ...               0.0           0.000000   
25%           1.180000  ...               0.0           0.000000   
50%           2.340000  ...               0.0           0.000000   
75%           2.990000  ...               0.0           1.000000   
max           3.960000  ...               0.0           1.000000   

       InternetService_bin  OnlineSecurity_bin  OnlineBackup_bin  \
count               7043.0         7043.000000       7043.000000   
mean                   0.0            0.286668          0.344881   
std                    0.0            0.452237          0.475363   
min                    0.0            0.000000          0.000000   
25%                    0.0            0.000000          0.000000   
50%                    0.0            0.000000          0.000000   
75%                    0.0            1.000000          1.000000   
max                    0.0            1.000000          1.000000   

       DeviceProtection_bin  TechSupport_bin  StreamingTV_bin  \
count           7043.000000      7043.000000      7043.000000   
mean               0.343888         0.290217         0.384353   
std                0.475038         0.453895         0.486477   
min                0.000000         0.000000         0.000000   
25%                0.000000         0.000000         0.000000   
50%                0.000000         0.000000         0.000000   
75%                1.000000         1.000000         1.000000   
max                1.000000         1.000000         1.000000   

       StreamingMovies_bin  servicios_contratados  
count          7043.000000            7043.000000  
mean              0.387903               2.459747  
std               0.487307               2.045539  
min               0.000000               0.000000  
25%               0.000000               1.000000  
50%               0.000000               2.000000  
75%               1.000000               4.000000  
max               1.000000               7.000000  

[8 rows x 21 columns]

In [49]:
datos.describe(include='O')

customerID gender MultipleLines InternetService OnlineSecurity  \
count         7043   7043          7043            7043           7043   
unique        7043      2             3               3              3   
top     0002-ORFBO   Male            No     Fiber optic             No   
freq             1   3555          3390            3096           3498   

       OnlineBackup DeviceProtection TechSupport StreamingTV StreamingMovies  \
count          7043             7043        7043        7043            7043   
unique            3                3           3           3               3   
top              No               No          No          No              No   
freq           3088             3095        3473        2810            2785   

              Contract     PaymentMethod  
count             7043              7043  
unique               3                 4  
top     Month-to-month  Electronic check  
freq              3875              2365

In [50]:
datos = datos.drop(columns='customerID')
datos.head()

Churn  gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0      0  Female              0        1           1       9             1   
1      0    Male              0        0           0       9             1   
2      1    Male              0        0           0       4             1   
3      1    Male              1        1           0      13             1   
4      1  Female              1        1           0       3             1   

  MultipleLines InternetService OnlineSecurity  ... PhoneService_bin  \
0            No             DSL             No  ...                0   
1           Yes             DSL             No  ...                0   
2            No     Fiber optic             No  ...                0   
3            No     Fiber optic             No  ...                0   
4            No     Fiber optic             No  ...                0   

  MultipleLines_bin InternetService_bin OnlineSecurity_bin OnlineBackup_bin  \
0                 0                   0                  0                1   
1                 1                   0                  0                0   
2                 0                   0                  0                0   
3                 0                   0                  0                1   
4                 0                   0                  0                0   

  DeviceProtection_bin  TechSupport_bin StreamingTV_bin  StreamingMovies_bin  \
0                    0                1               1                    0   
1                    0                0               0                    1   
2                    1                0               0                    0   
3                    1                0               1                    1   
4                    0                1               1                    0   

   servicios_contratados  
0                      3  
1                      2  
2                      1  
3                      4  
4                      2  

[5 rows x 32 columns]

# Encoding
> Transformación de variables categóricas a formato numérico

In [51]:
categoricas = [
    "gender",
    "MultipleLines",
    "InternetService",
    "OnlineSecurity",
    "OnlineBackup",
    "DeviceProtection",
    "TechSupport",
    "StreamingTV",
    "StreamingMovies",
    "Contract",
    "PaymentMethod",
]

datos_encoded = pd.get_dummies(data=datos, columns=categoricas, dtype=int)
datos_encoded.head()

Churn  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0      0              0        1           1       9             1   
1      0              0        0           0       9             1   
2      1              0        0           0       4             1   
3      1              1        1           0      13             1   
4      1              1        1           0       3             1   

   PaperlessBilling  Charges.Monthly  Charges.Total  Cuentas_Diarias  ...  \
0                 1             65.6         593.30             2.19  ...   
1                 0             59.9         542.40             2.00  ...   
2                 1             73.9         280.85             2.46  ...   
3                 1             98.0        1237.85             3.27  ...   
4                 1             83.9         267.40             2.80  ...   

   StreamingMovies_No  StreamingMovies_No internet service  \
0                   1                                    0   
1                   0                                    0   
2                   1                                    0   
3                   0                                    0   
4                   1                                    0   

   StreamingMovies_Yes  Contract_Month-to-month  Contract_One year  \
0                    0                        0                  1   
1                    1                        1                  0   
2                    0                        1                  0   
3                    1                        1                  0   
4                    0                        1                  0   

   Contract_Two year  PaymentMethod_Bank transfer (automatic)  \
0                  0                                        0   
1                  0                                        0   
2                  0                                        0   
3                  0                                        0   
4                  0                                        0   

   PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                      0                               0   
1                                      0                               0   
2                                      0                               1   
3                                      0                               1   
4                                      0                               0   

   PaymentMethod_Mailed check  
0                           1  
1                           1  
2                           0  
3                           0  
4                           1  

[5 rows x 54 columns]

# Verificar Proporción de Cancelación (Churn)

In [52]:
datos_encoded['Churn'].value_counts(normalize=True)*100

Churn
0    73.463013
1    26.536987
Name: proportion, dtype: float64

<p style='color: #E2E2E2; background-color: #1C0F13; padding: 24px'>A través del método anterior se puede observar que hay un gran desbalance en las clases analizadas. Se cree que esto podría impactar en los modelos predictivos</p>

# TODO: Balanceo de Clases

In [ ]:
# from imblearn.over_sampling import SMOTE

# oversampling = SMOTE()

# x_balanceada, y_balanceada = 

# Análisis de correlación

In [63]:
corr = datos_encoded.corr()
corr['Churn'].sort_values(ascending=False)*100

Churn                                      100.000000
Contract_Month-to-month                     40.510291
OnlineSecurity_No                           34.263680
TechSupport_No                              33.728073
InternetService_Fiber optic                 30.801974
PaymentMethod_Electronic check              30.191875
OnlineBackup_No                             26.800518
DeviceProtection_No                         25.248136
Cuentas_Diarias                             19.341174
Charges.Monthly                             19.335642
PaperlessBilling                            19.182533
SeniorCitizen                               15.088933
StreamingMovies_No                          13.084503
StreamingTV_No                              12.891588
StreamingTV_Yes                              6.322775
StreamingTV_bin                              6.322775
StreamingMovies_bin                          6.138155
StreamingMovies_Yes                          6.138155
MultipleLines_bin           

Las siguientes variables son candidatas para ser usadas en el modelo predictivo por su fuerte correlación con la cancelación (Churn), todas mayores al 25%:
- Contract_Month-to-month 40.51%
- OnlineSecurity_No 34.26%
- TechSupport_No 33.72%
- InternetService_Fiber optic 30.80%
- PaymentMethod_Electronic check 30.19
- OnlineBackup_No 26.80%
- DeviceProtection_No 25.24%

# Análisis dirigido de correlación

### Tiempo de contrato: `tenure`

In [65]:
datos_encoded[['Churn', 'tenure']].corr()

Churn    tenure
Churn   1.000000 -0.352229
tenure -0.352229  1.000000

In [68]:
import plotly.express as px
import matplotlib.pyplot as plt

In [75]:
fig = px.box(datos_encoded, x='Churn', y='tenure')

fig.show()

### Gasto total: `Charges.Total`

In [66]:
datos_encoded[["Churn", "Charges.Total"]].corr()

Churn  Charges.Total
Churn          1.000000      -0.198324
Charges.Total -0.198324       1.000000

In [76]:
fig = px.box(datos_encoded, x="Churn", y="Charges.Total")

fig.show()

En ambos casos se observa que hay una correlación negativa, más fuerte en la variable de tiempo de contratación.

Esto podría llegar a tener un sentido lógico.